Feature/exchange value estimation

sacOO7 opened the pull request:
- Created test orderbooks data for testing
- Added sample tests for both asks and bids.
- Refactored tests folder.

sacOO7 committed:
[created test-data orderbook raw fixtures for calculating estimated value
of returns corresponding to given asks or bids][references issue #3]

sacOO7 committed:
[Created first orderbook based on realtime queue data][references issue #3]

sacOO7 committed:
[added test for estimation based on orderbook][references issue #3]

sacOO7 committed:
[updated testes for lsh buyers and sellers][references issue #3]

sacOO7 committed:
[renamed test folder, making each test file to have .test postfix]
[references issue #3]

sacOO7 committed:
[modified test to represent domain driver scenarios][references issue #3]

sacOO7 committed:
[updated test orderbook, changed unit used to represent price]
[references issue #3]

sacOO7 committed:
[added utils to calculate estimated returns for buyer and seller]
[updated tests and scenarios for the same]
[references issue #3]

sacOO7 committed:
[renamed orderbook1 to orderbook][added manual test scenarios]
[references issue #3]

jondubois commented on lisk-dex-electron/src/Utils.js:
```diff
@@ -40,6 +40,54 @@ const Keys = (dict) => Object.keys(dict);
+  let status = EstimationStatus.NO_MATCH;
+  const maxAmountCanBeReturned = amount * price;
+  for (const bid of bids) {
+    const remainingAmount = maxAmountCanBeReturned - estimatedReturns;
```
Note that new orders should be given the **best possible deal** based on the current state of the order book.

For example, if the ask price which is passed to this function is 0.5 but the highest bid in the list is 0.7, then the amount taken from the first bid should be `amount * 0.7` (not `amount * 0.5`) - This is a higher amount.

Instead of relying on `maxAmountCanBeReturned` to convert the ask currency into the bid currency (`ask amount -> bid amount`), you should do the inverse conversion (`bid amount -> ask amount`) - That means you need to convert the base currency from each bid into the quote currency then subtract it from the `amount` (which is in the ask currency). You can use `bid.amount / bid.price` to do the conversion in reverse.

jondubois replied:
The other functions look good. We can remove both of the flat return ones and the related tests.

jondubois commented on lisk-dex-electron/src/Utils.js:
```diff
@@ -40,6 +40,54 @@ const Keys = (dict) => Object.keys(dict);
+const estimatedReturnsForBuyer = (amount, price, asks) => {
+  let estimatedReturns = 0;
+  let status = EstimationStatus.NO_MATCH;
+  const maxAmountCanBeReturned = amount * (1 / price);
```
This is a similar problem as with the `estimatedReturnsForSeller` function. It does not give the user the best deal - In this case the `best deal` means the lowest ask price. You should do the inverse conversion. Now it's doing `bid amount -> ask amount`, but the conversion should be done for each ask individually by converting `ask amount -> bid amount`.

jondubois reviewed (changes requested):
The price estimate should be based on the best possible deal so that it matches how the DEX works on the back end.

sacOO7 committed:
[Capped decimal places for estimated returns to 4 digits][references
issue #3]

sacOO7 committed:
[Fixed test based on partial matching][references issue #3]

sacOO7 committed:
[renamed existing return functions to flat returns]
[added more functions to calculate best returns]
[references issue #3]

sacOO7 committed:
[updated bestReturns utils and tests for the same][#3]

sacOO7 commented:
Thinking of adding one more status, UNFULLFILLED to denote, estimate reached end of total amount available. Currently it's available as PARTIAL_MATCH.

jondubois commented on lisk-dex-electron/src/Utils.js:
```diff
@@ -40,6 +40,100 @@ const Keys = (dict) => Object.keys(dict);
+        break;
+      } else {
+        estimatedReturns += bid.amount;
+        const amountSold = (bid.amount / bid.price);
```
No need for brackets around the division.

jondubois commented on lisk-dex-electron/src/Utils.js:
```diff
@@ -40,6 +40,100 @@ const Keys = (dict) => Object.keys(dict);
+const EstimationStatus = {
+  MATCH: 1,
+  PARTIAL_MATCH: 2,
+  NO_MATCH: 3,
```
Can be use strings here instead of integers? Makes it clearer during debugging.

sacOO7 committed:
[Removed redundant flat returns functions]
[Updated enum to support strings instead of numbers]
[Updated tests for the same]

sacOO7 committed:
[refactored orderbook][updated utils, added getUIOrderBook method]
[Added tests for the same][#3]

jondubois commented on lisk-dex-electron/src/Utils.js:
```diff
@@ -40,6 +40,84 @@ const Keys = (dict) => Object.keys(dict);
+};
+
+// todo - reuse the function in orderbook.js, make sure calculations are in one place
+const getUIOrderBookFrom = (contextOrderBook) => {
```
Can `getUIOrderBookFrom` be renamed to something like `getSanitizedOrderBook` or `cleanupRawOrderBookData` or something like that? Functions should ideally avoid making assumptions about how and where their return value is used; whether it's used by the UI or in the tests or elsewhere shouldn't matter.

sacOO7 replied:
I can go with sanitized, but word it too clean to be used.. I will just use getCleanOrderBook

jondubois commented on lisk-dex-electron/src/tests/Utils.test.js:
```diff
@@ -0,0 +1,117 @@
+    expect(groupByKey(actualArray, 'price', 'value')).toEqual(expectedReduction);
+  });
+
+  it('Gets keys from dict in correct order', () => {
```
There is no need to test the `Keys` function because what it does is very straight forward. Tests should aim to mitigate risk. This function is very low risk so doesn't need tests. Also, since this function is commonly used, it will be covered by other tests anyway.

Tests lock down the code base and make it difficult to adapt to changing requirements. We need to keep this flexibility at the moment.

jondubois commented on lisk-dex-electron/src/tests/Utils.test.js:
```diff
@@ -0,0 +1,117 @@
+    expect(Keys(actualDict)).toEqual(expectedKeys);
+  });
+
+  it('Gets values from dict in correct order', () => {
```
No need to test the `Values` function for the same reason as above.
We need to minimize the amount of tests. There is a cost to maintain them and they can discourage change. Change can be good.

sacOO7 committed:
[Refactored placeorder component]
[Added common utility function for calculating estimated returns based
on clean orderbook]
[removed redundunt tests for keys and values]
[#3]

jondubois commented:
Looks great thanks.
